rust/src/lib/ifaces/hsr.rs

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Serialize};
+
+use crate::{BaseInterface, InterfaceType, NmstateError};
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case")]
+#[non_exhaustive]
+/// HSR interface. The example YAML output of a
+/// [crate::NetworkState] with an HSR interface would be:
+/// ```yaml
+/// ---
+/// interfaces:
+///   - name: hsr0
+///     type: hsr
+///     state: up
+///     hsr:
+///       port1: eth1
+///       port2: eth2
+///       multicast-spec: 40
+///       protocol: prp
+/// ```
+pub struct HsrInterface {
+    #[serde(flatten)]
+    pub base: BaseInterface,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// Deserialize and serialize to `hsr`.
+    pub hsr: Option<HsrConfig>,
+}
+
+impl Default for HsrInterface {
+    fn default() -> Self {
+        Self {
+            base: BaseInterface {
+                iface_type: InterfaceType::Hsr,
+                ..Default::default()
+            },
+            hsr: None,
+        }
+    }
+}
+
+impl HsrInterface {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn sanitize(
+        &mut self,
+        is_desired: bool,
+    ) -> Result<(), NmstateError> {
+        if is_desired {
+            if let Some(conf) = &mut self.hsr {
+                if let Some(address) = &mut conf.supervision_address {
+                    address.as_mut().make_ascii_uppercase();
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case")]
+#[non_exhaustive]
+#[derive(Default)]
+pub struct HsrConfig {
+    /// The port1 interface name.
+    pub port1: String,
+    /// The port2 interface name.
+    pub port2: String,
+    /// The MAC address used for the supervision frames. This property is
+    /// read-only.
+    pub supervision_address: Option<String>,
```

Comment:
If this property is read only, we should raise an info message indicate we are ignoring desired supervision_address when user applying back the current state.

Reply:
Also need `#[serde(skip_serializing_if = "Option::is_none"`.